In [1]:
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt      
import math
import seaborn as sns
from collections import Counter
from sklearn.decomposition import PCA
import datetime 
import sklearn

%matplotlib inline

# sklearn imports
from sklearn import linear_model
from sklearn import metrics

# helper for logistic regression explanations
from scipy.special import expit
from scipy.special import logit
 
plt.rcParams['figure.figsize'] = (10, 6)

from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

# show several prints in one cell. This will allow us to condence every trick in one cell.
#from IPython.core.interactiveshell import InteractiveShell
#InteractiveShell.ast_node_interactivity = "all"

from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.linear_model import LinearRegression

from sklearn.metrics import confusion_matrix, classification_report, mean_absolute_error, r2_score

In [2]:
df = pd.read_csv('preply_newForq3.csv',sep=',')
df.drop('Unnamed: 0', inplace=True,axis=1)
df

price  lessons  rating  popular  has_diploma  approved_matirials  \
0        60      264     5.0        0            0                   0   
1       110     1724     5.0        0            0                   1   
2       126      663     5.0        1            1                   0   
3        28        3     0.0        1            1                   0   
4        28        2     0.0        3            1                   0   
...     ...      ...     ...      ...          ...                 ...   
4012     22        3     0.0        0            0                   0   
4013     47        2     0.0        0            0                   0   
4014     28        0     0.0        0            0                   0   
4015     28        0     0.0        0            0                   0   
4016     37       42     5.0        0            1                   0   

      reviews  students_count  English  teaches_count  last_booked  \
0          10              19        7              1            4   
1          25              28        7              1            1   
2          19              32        7              1            5   
3           0               5        7              1            2   
4           0               9        7              1           15   
...       ...             ...      ...            ...          ...   
4012        0               0        7              1            0   
4013        0               0        7              1            0   
4014        0               0        7              1            0   
4015        0               0        7              1            0   
4016        1               1        7              1            0   

      response_time  is_expensive  countSpeakLanguage  
0               1.0             0                   1  
1               1.0             1                   1  
2               7.0             1                   1  
3              24.0             0                   2  
4               4.0             0                   1  
...             ...           ...                 ...  
4012            1.0             0                   1  
4013            1.0             0                   1  
4014            1.0             0                   1  
4015            1.0             0                   1  
4016            1.0             0                   1  

[4017 rows x 14 columns]

# Research question: Is it possible to predict the number of lessons a teacher has?

In [3]:
df.lessons.describe()

count     4017.000000
mean       529.424197
std       1127.247924
min          0.000000
25%         10.000000
50%        116.000000
75%        529.000000
max      16307.000000
Name: lessons, dtype: float64

In [4]:
y = df['lessons']
X = df.drop('lessons', 1)

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [6]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [7]:
def describe_output_for_model(model, X_test, y_test, model_type):
    print("Predicting...")
    print("\tdisplaying information re: the '%s' model ...\n" %( model_type))
    y_pred = model.predict(X_test)
    if model_type == "classification":
        print("Model Accuracy: ", model.score(X_test, y_test))
        conf_mat = confusion_matrix(y_true=y_test, y_pred=y_pred)
        print(conf_mat)
        
        print("\nClassification Report: \n")
        print(classification_report(y_true=y_test, y_pred=y_pred))
    else:
        print("RMAE: ", mean_absolute_error(y_true=y_test, y_pred=y_pred))
        print("R^2 : ", r2_score(y_test, y_pred))
    
        
        
    # Coefficients
    if model_type == "classification":
        coef = list(sorted(zip(np.arange(X_test.shape[1]), abs(model.coef_[0])),key=lambda x: x[1]))
        coef_table = pd.DataFrame(np.array(coef).reshape(-1,2),columns=['Attributes','Coefficient'])
        print('\nCoefficient for every feature:\n')
        print(coef_table)
    else:
        coef = list(sorted(zip(np.arange(X_test.shape[1]), abs(model.coef_)),key=lambda x: x[1]))
        coef_table = pd.DataFrame(np.array(coef).reshape(-1,2),columns=['Attributes','Coefficient'])
        print('\nCoefficient for every feature:\n')
        print(coef_table)

describe_output_for_model(model, X_test, y_test, "regression")

Predicting...
	displaying information re: the 'regression' model ...

RMAE:  266.55702851129166
R^2 :  0.707966143989525

Coefficient for every feature:

    Attributes  Coefficient
0          0.0     2.346038
1         10.0     5.383066
2          9.0     9.369243
3         12.0    15.850200
4          3.0    20.192274
5          6.0    20.816736
6          1.0    21.346774
7          5.0    38.497370
8          7.0    72.718553
9          8.0    97.775541
10        11.0   111.722667
11         2.0   135.780093
12         4.0   254.448520


## As can be seen we got:
## a rate of 266.55702851129166 RMAE (Root mean square error)
## 0.707966143989525 rate of success (R^2 - goodness-of-fit measure for linear regression models)

